### ISSUE TITLE
[BUG] A CURL timeout when refreshing credentials causes subsequent putKinesisVideoFrame calls to fail silently

### ISSUE THREAD

**Issue by schultetwin1** [awslabs/amazon-kinesis-video-streams-producer-c#221]
**Media pipeline:**
 - What is the media source: Custom
 - Intermittent/continuous streaming: Intermittent

**Logging**

Here is the log we see. After the failure we only see calls to `putKinesisVideoFrame`. We never see `notifyDataAvailable`.

```
Jun 15 16:29:25 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997619962310, dts: 16237997619962310, duration: 0, size: 65183, trackId: 1, isKey 1
Jun 15 16:29:25 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997620015290, dts: 16237997620015290, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:25 kvs_stream: stepStateMachine(): State Machine - Current state: 0x0000000000000200, Next state: 0x0000000000000020
Jun 15 16:29:25 kvs_stream: stepStateMachine(): State Machine - Current state: 0x0000000000000040, Next state: 0x0000000000000002
Jun 15 16:29:40 kvs_stream: blockingCurlCall(): Curl perform failed for url https://*********.credentials.iot.us-west-2.amazonaws.com/role-aliases/****************/credentials with result Timeout was reached : Resolving timed out after 10000 milliseconds 
Jun 15 16:29:40 kvs_stream: putKinesisVideoFrame(): Failed to submit frame to Kinesis Video client. status: 0x15000011 decoding timestamp: 16237997620015290 presentation timestamp: 16237997620015290
Jun 15 16:29:40 kvs_stream: [kvs_module.c:700]ret=0x15000011 [700]seq=258717 frame size=320 time=1623799762001529
Jun 15 16:29:40 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997620391790, dts: 16237997620391790, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997620650150, dts: 16237997620650150, duration: 0, size: 7009, trackId: 1, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997620795270, dts: 16237997620795270, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997621191110, dts: 16237997621191110, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997621277220, dts: 16237997621277220, duration: 0, size: 2161, trackId: 1, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997621587740, dts: 16237997621587740, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997621976830, dts: 16237997621976830, duration: 0, size: 8807, trackId: 1, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997621990100, dts: 16237997621990100, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997622388610, dts: 16237997622388610, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997622632750, dts: 16237997622632750, duration: 0, size: 7024, trackId: 1, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997622788910, dts: 16237997622788910, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997623056530, dts: 16237997623056530, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997623317530, dts: 16237997623317530, duration: 0, size: 6927, trackId: 1, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997623589650, dts: 16237997623589650, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997623944730, dts: 16237997623944730, duration: 0, size: 7069, trackId: 1, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997623989540, dts: 16237997623989540, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997624401270, dts: 16237997624401270, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997624659060, dts: 16237997624659060, duration: 0, size: 2506, trackId: 1, isKey 0
Jun 15 16:29:41 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237997624791860, dts: 16237997624791860, duration: 0, size: 320, trackId: 2, isKey 0
...
```

Here is what I would expect to see:

```
Jun 15 16:01:47 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237981055322370, dts: 16237981055322370, duration: 0, size: 54335, trackId: 1, isKey 1
Jun 15 16:01:47 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237981055530400, dts: 16237981055530400, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:01:47 kvs_stream: stepStateMachine(): State Machine - Current state: 0x0000000000000200, Next state: 0x0000000000000040
Jun 15 16:01:47 kvs_stream: [kvs_module.c:178]Stream Ready. Stream Handle: 5553104
Jun 15 16:01:47 kvs_stream: ret = 0 0 117
Jun 15 16:01:47 kvs_stream: stepStateMachine(): State Machine - Current state: 0x0000000000000040, Next state: 0x0000000000000080
Jun 15 16:01:47 kvs_stream: putStreamResultEvent(): Put stream result event. New upload handle 2
Jun 15 16:01:47 kvs_stream: notifyDataAvailable(): Note data received: duration(100ns): 0 bytes 0 for stream handle 2
Jun 15 16:01:47 kvs_stream: notifyDataAvailable(): Failed to un-pause curl with error: 43. Curl object 0x5701c8
Jun 15 16:01:47 kvs_stream: defaultStreamDataAvailableCallback(): Reported streamDataAvailable callback for stream handle 5553104. Upload handle 2. Duration available in 100ns: 0. Bytes available: 0
Jun 15 16:01:47 kvs_stream: stepStateMachine(): State Machine - Current state: 0x0000000000000080, Next state: 0x0000000000000100
Jun 15 16:01:47 kvs_stream: notifyDataAvailable(): Note data received: duration(100ns): 0 bytes 54731 for stream handle 2
Jun 15 16:01:47 kvs_stream: defaultStreamDataAvailableCallback(): Reported streamDataAvailable callback for stream handle 5553104. Upload handle 2. Duration available in 100ns: 0. Bytes available: 54731
Jun 15 16:01:47 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237981055928860, dts: 16237981055928860, duration: 0, size: 320, trackId: 2, isKey 0
Jun 15 16:01:47 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237981056204300, dts: 16237981056204300, duration: 0, size: 4078, trackId: 1, isKey 0
Jun 15 16:01:47 kvs_stream: notifyDataAvailable(): Note data received: duration(100ns): 210000 bytes 55064 for stream handle 2
Jun 15 16:01:47 kvs_stream: defaultStreamDataAvailableCallback(): Reported streamDataAvailable callback for stream handle 5553104. Upload handle 2. Duration available in 100ns: 210000. Bytes available: 55064
Jun 15 16:01:47 kvs_stream: notifyDataAvailable(): Note data received: duration(100ns): 600000 bytes 55397 for stream handle 2
Jun 15 16:01:47 kvs_stream: defaultStreamDataAvailableCallback(): Reported streamDataAvailable callback for stream handle 5553104. Upload handle 2. Duration available in 100ns: 600000. Bytes available: 55397
Jun 15 16:01:47 kvs_stream: putKinesisVideoFrame(): debug frame info pts: 16237981056333450, dts: 16237981056333450, duration: 0, size: 320, trackId: 2, isKey 0
...
```

**Describe the bug**
It appears that the SDK will automatically refresh credentials every hour. If that credential refresh fails due to a libcurl timeout, all subsequent calls to `putKinesisVideoFrame()` will fail silently. The return code of `putKinesisVideoFrame()` will be `STATUS_SUCCESS` but the internal state machine of the KVS SDK will not actually upload any frames.

**SDK version number**
1.1.1

**To Reproduce**
1. Start an intermittent producer
2. Successfully stream data to KVS servers
3. Stop producing frames
3. Wait 1 hour
4. Purposefully degrade the network connection.
5. Start producing frames using `putKinesisVideoFrame`
6. Wait for log to show `putKinesisVideoFrame()` failed with status `0x15000011`
7. Remove network degradation.
8. Notice that `putKinesisVideoFrame` is now returning "success" but frames are not being uploaded (no ACK's are being received).

**Expected behavior**
When the network degradation is removed, I would expect to start getting acks again from the AWS KVS servers.

**Desktop (please complete the following information):**
 - OS: Buildroot Linux
 
**Additional context**
Internal ticket number is 8260728551.

**Comment by hassanctech** [comment]
Have a partial root cause here so basically the stream state machine when going from stopped --> get endpoint state it steps the client state machine from stopped --> auth, when this auth fails the bool that was set in `checkStreamingTokenExpiration`, specifically `gracePeriod` is not reset so all subsequent `putKinesisVideoFrames` short circuit here and return no error from `checkStreamingTokenExpiration`.  To resolve this I intend to reset the `gracePeriod` if stepping the client state machine failed, then if all retries on the client state machine are exhausted I will transition it back to READY state.  The stream state machine already has in place when `putKinesisVideoFrame` is called when it invokes `streamTerminatedEvent` the supplied parameters will always transition it back to `STREAM_STATE_STOPPED` and then to `STREAM_STATE_GET_ENDPOINT`

**Comment by schultetwin1** [comment]
Thank you!

Let me know when you are ready for testing. I can reliably reproduce this error, so can verify the fix actually fixes it.

**Comment by hassanctech** [comment]
Ok great!  You can test what's in the PR right now -- it won't be the final version I'll be making some improvements as suggested by @MushMal's comments and should have another revision out but what's there should work. If it's easy to test, or else you can just wait for the final version I'll update here when that's ready!

**Comment by hassanctech** [comment]
@schultetwin1 if you do try it please let me know if it resolved the issue.

**Comment by hassanctech** [comment]
@schultetwin1 https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/master/src/source/Common/IotCredentialProvider.c#L182

This is what sets the interval at which the creds will expire and we'll try to refresh auth, the value of MAX_ENFORCED_TOKEN_EXPIRATION_DURATION is 40 minutes, defined here:

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/9ef93e5a41465ec0c847b8e1ee30ac9ba29fdb7f/src/client/include/com/amazonaws/kinesis/video/client/Include.h#L290

Basically we make the auth back-end call which tell us when creds expire and we set the time to the shorter of that time or current time + 40 mins.  If you change `MAX_ENFORCED_TOKEN_EXPIRATION_DURATION` to say 2 minutes you can test much quicker next time. In any case I'm setting up tests on my end as well so I can reproduce this locally and see what's going on.

**Comment by schultetwin1** [comment]
@hassanctech - Feel free to close this issue as it's now fixed.
